src/de.rs

```diff
@@ -358,10 +358,43 @@ impl<'de, R: dec::Read<'de>> serde::Deserializer<'de> for &mut Deserializer<R> {
         u64,        deserialize_u64,        visit_u64;
         u128,       deserialize_u128,       visit_u128;
 
-        f32,        deserialize_f32,        visit_f32;
+        // f32 deserialize is handled as a special case below
         f64,        deserialize_f64,        visit_f64;
     );
 
+    #[inline]
+    fn deserialize_f32<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        // DAG-CBOR strictly requires all floats to be encoded as f64.
+        // When deserializing to f32, we need to handle f64 encoding.
+        // We also accept f32 encoding, although a strict DAG-CBOR
+        // implementation should reject it (please don't write new data
+        // with f32 encoding).
+        let byte = peek_one(&mut self.reader)?;
+        match byte {
+            marker::F32 => {
+                // Note: f32 encoding is not valid in strict DAG-CBOR.
+                let value = <f32>::decode(&mut self.reader)?;
+                visitor.visit_f32(value)
+            }
+            marker::F64 => {
+                // DAG-CBOR always uses f64 encoding, even for f32 values
+                let value = <f64>::decode(&mut self.reader)?;
+                let f32_value = value as f32;
```

Comment:
One more idea: Should we only convert to f32 if it fits in, i.e. it was a round-trip? We could add a check that converts `f32_value` to an `f64` and check if the bits are the same.

Reply:
I think we might bump into the `0!=-0` case -- `0x00000000` vs `0x80000000` just by value checking.